apply.json

```
@@ -16,7 +16,7 @@
         },
         {
             "name": "process",
-            "description": "A unary process to be applied on each value, may consist of multiple sub-processes.",
+            "description": "A process to be applied on each individual value, may consist of multiple sub-processes. The process must accept and return a single value, for example ``abs()`` or ``linear_scale_range()``.",
```

`linear_scale_range`  requires at least 3 arguments, so that would be a wrong example, right? I understand that you can "hardcode" the `inputMin` and `inputMax` in the sub-process-graph, but I think  the current description is pretty confusing for the less-experienced openEO user.

---

It's exactly there for this purpose, to show that it is about the "data" values (parameter "x"). The others are basically the "options". Hard to describe, but the aim was to avoid that people only look for processes with exactly one parameter. I'm not exactly sure how to improve the documentation right now.

---

I don't think there is "room" to explain how to use `linear_scale_range` in `apply` (except maybe with an actual example), so I would keep it superficial with indeed something like `ln`, `log`, `sin`, ...

---

log and linear_scale_range have the same "issue" with requiring an additional parameter. I don't see a difference, to be honest.

---

ah ok, I was assuming that `log` had default `base=10`, but apparently not, so `log` has indeed same issue.

My original remark is that current description is misleading on several levels:

> The process must accept and return a single value, for example ``abs()`` or ``linear_scale_range()``

Superficially: `linear_scale_range` does not accept a single value, it has three required arguments, so this statement looks wrong.

More precisely:  you can not pass `abs` or `linear_scale_range` directly to `apply`, you have to embed it in a process graph (which as a whole indeed  "must accept and return a single value"), but that statement does not reflect that either.

---

> Superficially: `linear_scale_range` does not accept a single value, it has three required arguments, so this statement looks wrong.

Yes, but as it is mentioned users will acknowledge that this is allowed. Just not mentioning it will make users wonder whether it's allowed to use such processes or not. So I think it's better to list it than not to list it. I agree that it's not ideal though.

> More precisely: you can not pass `abs` or `linear_scale_range` directly to `apply`, you have to embed it in a process graph (which as a whole indeed "must accept and return a single value"), but that statement does not reflect that either.

The process descriptions are aimed towards users, which will use client libraries. There, process graphs are hidden and you can actually directly pass those functions. To improve the situation a bit:

```suggestion
            "description": "A process that accepts and returns a single value and is applied on each individual value in the data cube. The process may consist of multiple sub-processes and could, for example, consist of the processes ``abs()`` or ``linear_scale_range()``.",
```

---

done.